std-lib/list/list.lurk

```diff
@@ -0,0 +1,157 @@
+  (current-env))
+
+;; list-n : make an empty list of size n
+(letrec 
+    ((list-n-iter (lambda (n list)
+        (if (= n 0)
+            list
+            (list-n-iter (- n 1) (cons nil list)))))
+    (list-n (lambda (n) (list-n-iter n nil))))
+    (current-env))
+
+;; numbered-list-n : make a numbered list from 0 to n of length (n + 1)
+(letrec 
+    ((numbered-list-iter (lambda (n list) 
+        (if (= n 0)
+            (cons 0 list)
+            (numbered-list-iter (- n 1) (cons n list)))))
+    (numbered-list-n (lambda (n)
+        (numbered-list-iter n nil))))
+    (current-env))
+
+
+;; nth : get the nth (0-indexed) value of a list 
+(letrec ((nth (lambda (place list)
+    (if (car list)
+        (if (= 0 place)
+            (car list)
+            (nth (- place 1) (cdr list)))))))
+    (current-env))
+
+;; last : get the last element of a list
+(let ((last (lambda (list) (car (reverse list))))) (current-env))
+
+;; list-update : update the nth (0-indexed) value of a list 
+;;               if (< (length list) n) then list-update returns 
+;;               an unchanged list
+(letrec 
+    ((list-update-iter (lambda (place value old-list new-list)
+        (if (car old-list)
+            (if (= 0 place)
+                (list-update-iter 
+                    (- place 1) 
+                    nil 
+                    (cdr old-list) 
+                    (snoc value new-list))
+                (list-update-iter
+                    (- place 1)
+                    value 
+                    (cdr old-list)
+                    (snoc (car old-list) new-list)))
+            new-list)))
+    (list-update (lambda (place value list) 
+        (list-update-iter place value list nil)))) 
+    (current-env))
+
+;; length : determines the (0-indexed) length of a list 
+(letrec 
+    ((length-iter (lambda (lst cntr)
+		    (if lst
+			(length-iter (cdr lst) (+ 1 cntr))
+			cntr)))
+     (length (lambda (lst) (length-iter lst 0))))
+  (current-env))
+
+;; apply : a preliminary version of apply
+(letrec 
+    ((apply1 (lambda (f list)
+               (if list
+		   (if (cdr list)
+                       (apply1 (f (car list)) (cdr list))
+                       (f (car list)))
+		   (f)))))
+  (current-env))
+
+;; map : a list map function 
+;; TODO : map that can take multiple lists
+(letrec ((map (lambda (f list)
+                    (if (eq list nil)
+                        nil
+                        (cons (f (car list))
+                              (map f (cdr list)))))))
+      (current-env))
+
+;; filter : takes a predicate and a list and returns only the items
+;; that match the predicate
+(letrec ((filter (lambda (pred lst)
+                   (if lst
+                       (if (pred (car lst))
+                           (cons (car lst) (filter pred (cdr lst)))
+                           (filter pred (cdr lst)))))))
+  (current-env))
+
+;; fold : a fold function over lists 
+(letrec 
+    ((fold (lambda (op accum lst)
```

Comment:
For stylistic, consistency, let's use ACC for accumulator variable. I'd personally use L for lists also, but I feel less strongly about that. LST doesn't feel very Lispy. LIST would be better, but more verbose (hence my preference for L).

Reply:
I don't like `list` because it collides with the function `list`. Although that brings up the question of whether lurk is a lisp-1 or lisp-2. let's see:

```lisp
> (let ((a (lambda () 2))
	(a 1))
    (a))
[6 iterations] => ERROR!
```

lisp-1. So naming collision is a problem for LIST. I agree with L then.